ceph_iscsi_config/lun.py

```diff
@@ -1231,38 +1231,41 @@ def define_luns(logger, config, target):
-                                                     "with LIO: {}"
-                                                     .format(disk_key,
-                                                             lun.error_msg))
-
-                            # If not in use by another target on this gw
-                            # clean up stale locks.
-                            if so.status != 'activated':
-                                RBDDev.rbd_lock_cleanup(logger, ips,
-                                                        rbd_image)
-
-                            target._map_lun(config, so, target_disks[disk_key])
-                            if target.error:
-                                raise CephiSCSIError("Mapping for {} failed: {}"
-                                                     .format(disk_key,
-                                                             target.error_msg))
+                        try:
+                            with rbd.Image(ioctx, image_name) as rbd_image:
+
+                                disk_config = config.config['disks'][disk_key]
+                                backstore = disk_config['backstore']
+                                backstore_object_name = disk_config['backstore_object_name']
+
+                                lun = LUN(logger, pool, image_name,
+                                          rbd_image.size(), local_gw, backstore,
+                                          backstore_object_name)
+
+                                if lun.error:
+                                    raise CephiSCSIError("Error defining rbd image {}"
+                                                         .format(disk_key))
+
+                                so = lun.allocate()
+                                if lun.error:
+                                    raise CephiSCSIError("Unable to register {} "
+                                                         "with LIO: {}"
+                                                         .format(disk_key,
+                                                                 lun.error_msg))
+
+                                # If not in use by another target on this gw
+                                # clean up stale locks.
+                                if so.status != 'activated':
+                                    RBDDev.rbd_lock_cleanup(logger, ips,
+                                                            rbd_image)
+
+                                target._map_lun(config, so, target_disks[disk_key])
+                                if target.error:
+                                    raise CephiSCSIError("Mapping for {} failed: {}"
+                                                         .format(disk_key,
+                                                                 target.error_msg))
+                        except rbd.ImageNotFound:
+                            logger.debug("Rbd image not found: '{}' ".format(image_name))
 
```

Comment:
I don't know the answer to this, maybe someone more experienced with ceph-iscsi and rbd. From what it looks like. It's grabbing the list of disks from the ceph-iscsi config, and then opening up the pool and searching for the key/object (disk) in ceph. So the question is what is the expected behaviour if the image in the pool is gone?

Do we just log to debug? that seems like it's easily missed, or do be need to do something more noticeable for ops?

I'd assume we'd either have to use:
  `logger.warning`

So it appears in more logs... and might even be colour coded depending on the log analyser.

Or maybe it should throw a CephISCSIError, but I assume that'll stop processing and not actually fix things, minus stop a full blown stacktrace.

Reply:
The original intention was to fail the service so that it could easily be seen that somehow the configuration is invalid and needs to be resolved. Just hiding it like this will still potentially leave dangling LUNs in the configuration. In this case, the easy "solution" is to just recreate the missing images and remove them from Ceph iSCSI first.

Reply:
@matthewoliver  @dillaman  Thanks for clarifying.
Cloud you point me how to recreate missing images when the rbd-target-api is failed? 
do you mean that we need to use cephcli to create a rbd image as the ceph-iscsi service is dead.

Reply:
Yeah, just use `rbd create --size 1G (??) --image-format exclusive-lock volumes/<image name>`